### PR TITLE
Quick Tune: PID 'auto manual' tune and oscillation detection

### PR DESCRIPTION
**Pull request by IamPete1**
This is a auto oscillation detection algorithm that can scale back the gains to stop oscillation. Needs testing on more vehicles and in real life. This is a demo:

https://youtu.be/goSFD2A3qfM

This is a vid that explains the methodology before I did the  AP implementation.    

https://youtu.be/3Q1a2skm17g

I suspect this will probability be fine to leave running all the time on most vehicles (currently there is no way to switch it off!), the issue is more likely that it is not sensitive enough rather than the other way. There is quite a lot that can be done to change the sensitivity with the various filter values and thresholds, I have not found a good way to pick them yet tho.

In testing I just ran three at once on each aixs with a different filter frequency, this worked quite well however its hard to say if its worth the extra cpu/memory cost. Again we can do a better job using nicer filters, but this requires a keeping  buffers of past values. 

The next step will be a [Ziegler Nichols](https://en.wikipedia.org/wiki/Ziegler%E2%80%93Nichols_method) based 'quick tune'.

### CODE REVIEW

**Comment by bnsgeyer** [comment]
@IamPete1 Nice work!  I am working on autotune for tradheli.  My approach is frequency based and I have developed a method to determine the frequency response of the vehicle and from that I am able to determine the Rate P and Rate D gains that would cause instability.  You can read more on my Posts in the [Tradheli Autotune thread](https://discuss.ardupilot.org/t/is-autotune-for-traditional-helicopter-in-sight/54619/40).  It would be great to talk with you about this more.

**Comment by IamPete1** [comment]
@bnsgeyer Interesting, it looks to me that our amplitude ratio's are the same calculation more or less. This code just runs it the whole time rather than having to do a set input. I think I roughly understand how you get the gains. 

For my  Ziegler Nichols I was planning to use P only, turn it up until the detector kicks in, this gives the critical P. Then use zero crossing to get the critical frequency. 

Another method would be to do a 'manual' type tune, turn up P then D then P again until the detector kicks in, then apply some safety factor. 

Not sure of the best way to tune FF, I think that is important for heli in particulate, and ideas?

The idea here is to use the same method to tune all the PID controllers, not a good as a method dedicate to a particular vehicle, but hopefully will give a good starting point.

**Comment by bnsgeyer** [comment]
@IamPete1 the technique I am using will work for  any vehicle even if it doesn't use the FF gain.  It is doing a stability analysis as one would do with a frequency response of a vehicle except it is identifying the frequency response data with a series of dwells. With this method, there is no need to raise the gain to the point where the vehicle oscillates. The series of dwells are conducted with a gain that doesn't cause oscillation. Then the ultimate gain can you calculated. 
The reference for the Ziegler-Nichols method will be very helpful in determining the knock down factor for the gains. I need to research this more.

**Comment by lthall** [comment]
Hi Pete, Nice work.

With multirotors the rule I follow is that if I reduce D then I reduce P and I by the same percentage. If you do not do this the reduced D term can result in the P term oscillating. On a well tuned aircraft this is almost always the case.

So I would suggest:
- if you detect oscillation in D then reduce all three PID terms.
- if you detect oscillation in P or I you can reduce them independently.

However, on a multirotor or heli it may be worth simply running this on the PID output and reducing all by the same amount.

Another issue that you need to be mindful of. On a well tuned aircraft with a high Angle_P value you risk starting an angular oscillation if you reduce P, I or D to stop a angular rate oscillation.

The simplest solution to the problems above would be to run this test on the rate output and target rate input.
- rate output oscillation would result in a multiplier to the whole PID output and target rate input. This ensures that the angle loop stays stable with lower PID gain.
- rate target oscillation would result in a multiplier to the rate target only.

So just some food for thought.

A heads up on  Ziegler Nichols method. It is well recognised that it works for only a very limited number of systems (and not necessarily well). It has also been the goto for multirotor autotuning attempts for a long time. It has never worked.

**Comment by IamPete1** [comment]
@lthall  It sounds like you don't think there is so much benefit in being able to tell witch term is causing the oscillation? Better to just reduce all. That would cut down the memory costs, we could run more filter frequencys and maybe do a better job at at spotting the oscillations in the first place. 

I think I will do a 'pseudo manual' tuner if you think there is not much benefit in Ziegler Nichols, this also means there is no need to do zero crossing detection to get the osculation frequency.

**Comment by IamPete1** [comment]
https://youtu.be/YbOZWb8pddk

**Comment by IamPete1** [comment]
I am now happy with this feature wise, needs testing on all vehicles. My plan now is to get this passing checks and look to get it in for testing. To get this working well on a range of vehicles I suspect some messing will have to be done with filter values.

Currently this has no consent of time, just counts calls. May need to add that to make it more robust.

File: libraries/APM_Control/AP_PitchController.h [hunk 1]
```diff
@@ -48,6 +50,8 @@ class AP_PitchController {
     AP_Float &kD(void) { return gains.D; }
     AP_Float &kFF(void) { return gains.FF; }
 
+    AC_PID_OscillationDetector oscillationDetector{_pid_info, gains.P, gains.I, gains.D, gains.FF, 2.0f, 2.0f, 0.01f, 1.0f}; // 2 cdeg/s magnitude threshold, 2 oscillation threshold, 0.01 filter value
```
Comment (IamPete1) [inline libraries/APM_Control/AP_PitchController.h:53]: use 1:5  I to P ratio

File: libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp [hunk 2]
```diff
@@ -217,9 +221,9 @@ const AP_Param::GroupInfo AC_AttitudeControl_Multi::var_info[] = {
...
-    _pid_rate_yaw(AC_ATC_MULTI_RATE_YAW_P, AC_ATC_MULTI_RATE_YAW_I, AC_ATC_MULTI_RATE_YAW_D, 0.0f, AC_ATC_MULTI_RATE_YAW_IMAX, AC_ATC_MULTI_RATE_RP_FILT_HZ, AC_ATC_MULTI_RATE_YAW_FILT_HZ, 0.0f, dt)
+    _pid_rate_roll(AC_ATC_MULTI_RATE_RP_P, AC_ATC_MULTI_RATE_RP_I, AC_ATC_MULTI_RATE_RP_D, 0.0f, AC_ATC_MULTI_RATE_RP_IMAX, AC_ATC_MULTI_RATE_RP_FILT_HZ, 0.0f, AC_ATC_MULTI_RATE_RP_FILT_HZ, dt, oscillation_detector_magnitude_threshold, oscillation_detector_threshold, oscillation_detector_filter),
+    _pid_rate_pitch(AC_ATC_MULTI_RATE_RP_P, AC_ATC_MULTI_RATE_RP_I, AC_ATC_MULTI_RATE_RP_D, 0.0f, AC_ATC_MULTI_RATE_RP_IMAX, AC_ATC_MULTI_RATE_RP_FILT_HZ, 0.0f, AC_ATC_MULTI_RATE_RP_FILT_HZ, dt, oscillation_detector_magnitude_threshold, oscillation_detector_threshold, oscillation_detector_filter),
+    _pid_rate_yaw(AC_ATC_MULTI_RATE_YAW_P, AC_ATC_MULTI_RATE_YAW_I, AC_ATC_MULTI_RATE_YAW_D, 0.0f, AC_ATC_MULTI_RATE_YAW_IMAX, AC_ATC_MULTI_RATE_RP_FILT_HZ, AC_ATC_MULTI_RATE_YAW_FILT_HZ, 0.0f, dt, oscillation_detector_magnitude_threshold, oscillation_detector_threshold, oscillation_detector_filter)
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp:226]: set PI ratio of 0.1

File: libraries/APM_Control/AP_RollController.h [hunk 3]
```diff
@@ -55,6 +56,8 @@ class AP_RollController {
     AP_Float &kD(void) { return gains.D; }
     AP_Float &kFF(void) { return gains.FF; }
 
+    AC_PID_OscillationDetector oscillationDetector{_pid_info, gains.P, gains.I, gains.D, gains.FF, 1.0f, 2.0f, 0.01f, 1.0f}; // 1 cdeg/s magnitude threshold, 2 oscillation threshold, 0.01 filter value
```
Comment (tridge) [inline libraries/APM_Control/AP_RollController.h:59]: plane roll/pitch ratio P:I should be 5:1 approx

File: libraries/AC_PID/AC_PID_OscillationDetector.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+{
+    // apply filter to P, I, D, actual and target
+    P_filt = P_filt * (1-w) + pid.P * w;
+    I_filt = I_filt * (1-w) + pid.I * w;
+    D_filt = D_filt * (1-w) + pid.D * w;
+    ff_filt = ff_filt * (1-w) + pid.FF * w;
+    target_filt = target_filt * (1-w) + pid.target * w;
+    actual_filt = actual_filt * (1-w) + pid.actual * w;
+
+    // Caulate the filtered envelope of the difference between the filter vale and actual
```
Comment (tridge) [inline libraries/AC_PID/AC_PID_OscillationDetector.cpp:52]: abs() - guess who uses matlab ?

**Review by tridge — CHANGES_REQUESTED** [review]
flash size?
cpu cost?

**Comment by IamPete1** [comment]
Master Plane CubeBlack

bin/arduplane  1444852  1468  195348  1641668

This PR

bin/arduplane  1450960  1468  195344  1647772

so + 6104, quite a large increase, plane is the worst case, adding 5 instances. So ~1220 per instance.

(also rebased)

I think there is room for savings, we could drop doing the terms individually, that is off limited value in anycase.

**Comment by IamPete1** [comment]
Dropping individual term filters reduces the size less than expected. Now this cant tell what is causing the oscillation it just reduces everything equally. 

bin/arduplane  1450516  1468  195348  1647332

so + 5664 over master, ~ 1132.8 per instance.

**Comment by amilcarlucas** [comment]
Needs a rebase :(

**Comment by IamPete1** [comment]
@amilcarlucas this is waiting on the plane to move to AC_PID, that will make the patch significantly smaller. https://github.com/ArduPilot/ardupilot/pull/15377

**Comment by amilcarlucas** [comment]
@IamPete1 Thanks for the clarification.

**Comment by IamPete1** [comment]
Closing this on the bases that the new auto tune for Plane is a better base for porting auto tune over to other vehicles.

**Pull request closed without merging**